### DISCUSSION TITLE
How to retrieve firebase user displayName by authorId(uid) that stored in firestore

### DISCUSSION THREAD

**Discussion by thanhtutzaw** [vercel/next.js#52380] [Help] [answered]
### Summary

I have post array that stored in firestore  with SSR . It includes authorID (uid).
Then I need to display author name by uid . I have tried firebase admin SDK in SSR code . 
But I got error when getUserData function is inside post.map()
error ./node_modules/@google-cloud/storage/build/src/bucket.js:21:0
Module not found: Can't resolve 'fs'
https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/@google-cloud/storage/build/src/index.js
./node_modules/firebase-admin/lib/storage/storage.js
./node_modules/firebase-admin/lib/app/firebase-namespace.js
./node_modules/firebase-admin/lib/default-namespace.js
./node_modules/firebase-admin/lib/index.js
./lib/firebaseAdmin.ts
./lib/firebase.ts
./pages/_app.tsx

**Comment by gerardnico** [comment]
Your error is `Module not found: Can't resolve 'fs'`

This module is a Node module. Ie the [file system module](https://nodejs.org/api/fs.html).

It seems then that [_app.tsx](https://nextjs.org/docs/pages/building-your-application/routing/custom-app) is running in the client (browser) and not in the server (ie nodejs).

If you want to use a Node package, you need to run your code on the server. The [api routes](https://nextjs.org/docs/pages/building-your-application/routing/api-routes) do that for instance. Write your code there.

**Reply by thanhtutzaw** [reply]
Both _app.tsx and index.tsx are using getServerSideProps. I only get error when I write getUserData function inside post.map().

**Comment by thanhtutzaw** [comment, marked as answer]
This problem came when I call it in postToJSON . All I need to do is call it after postToJSON .  I don't know why but It should also work in postToJSON fuction .

### Before

```
const posts = await Promise.all(
      postSnap.docs.map(async (doc) => {
        const post = await postToJSON(doc); // userToJSON called in postToJSON
        return {
          ...post,
        };
      })
   );
```
 
    
### After

```
const posts = await Promise.all(
      postSnap.docs.map(async (doc) => {
        const post = await postToJSON(doc);
        const UserRecord = (await getUserData(post.authorId)) as UserRecord;
        const userJSON = userToJSON(UserRecord);
        return {
          ...post,
          author: {
            ...userJSON,
          },
        };
      })
    );
```
